model/textparse/nhcbparse.go

```diff
@@ -0,0 +1,344 @@
+// Copyright 2024 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package textparse
+
+import (
+	"errors"
+	"io"
+	"math"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/util/convertnhcb"
+)
+
+// The NHCBParser wraps a Parser and converts classic histograms to native
+// histograms with custom buckets.
+//
+// Since Parser interface is line based, this parser needs to keep track
+// of the last classic histogram series it saw to collate them into a
+// single native histogram.
+//
+// Note:
+// - Only series that have the histogram metadata type are considered for
+//   conversion.
+// - The classic series are also returned if keepClassicHistograms is true.
+type NHCBParser struct {
+	// The parser we're wrapping.
+	parser Parser
+	// Option to keep classic histograms along with converted histograms.
+	keepClassicHistograms bool
+
+	// Labels builder.
+	builder labels.ScratchBuilder
+
+	// Caches the values from the underlying parser.
+	// For Series and Histogram.
+	bytes []byte
+	ts    *int64
+	value float64
+	h     *histogram.Histogram
+	fh    *histogram.FloatHistogram
+	// For Metric.
+	lset         labels.Labels
+	metricString string
+	// For Type.
+	bName []byte
+	typ   model.MetricType
+
+	// Caches the entry itself if we are inserting a converted NHCB
+	// halfway through.
+	entry            Entry
+	err              error
+	justInsertedNHCB bool
```

Comment:
I have an instinct that we could avoid so many bools around like `justInsertedNHCB`, `isCollationInProgress`, `lastHistogramName==""` with one bool or enum -- might improve readability here.

But it's good enough for now and I didn't test this theory, so let's merge and leave that as an idea for the future (: